xpra/server/auth/keycloak_auth.py

```diff
@@ -0,0 +1,125 @@
+KEYCLOAK_REALM_NAME = os.environ.get("XPRA_KEYCLOAK_REALM_NAME", "example_realm")
+KEYCLOAK_CLIENT_ID  = os.environ.get("XPRA_KEYCLOAK_CLIENT_ID", "example_client")
+KEYCLOAK_CLIENT_SECRET_KEY = os.environ.get("XPRA_KEYCLOAK_CLIENT_SECRET_KEY", "secret")
+KEYCLOAK_REDIRECT_URI = os.environ.get("XPRA_KEYCLOAK_REDIRECT_URI", "http://localhost/login/")
+KEYCLOAK_SCOPE = os.environ.get("XPRA_KEYCLOAK_SCOPE", "openid")
+KEYCLOAK_GRANT_TYPE = os.environ.get("XPRA_KEYCLOAK_GRANT_TYPE", "authorization_code")
+
+
+class Authenticator(SysAuthenticator):
+
+    def __init__(self, **kwargs):
+        self.server_url = kwargs.pop("server_url", KEYCLOAK_SERVER_URL)
+        self.realm_name = kwargs.pop("realm_name", KEYCLOAK_REALM_NAME)
+        self.client_id = kwargs.pop("client_id", KEYCLOAK_CLIENT_ID)
+        self.client_secret_key = kwargs.pop("client_secret_key", KEYCLOAK_CLIENT_SECRET_KEY)
+        self.redirect_uri = kwargs.pop("redirect_uri", KEYCLOAK_REDIRECT_URI)
+        self.scope = kwargs.pop("scope", KEYCLOAK_SCOPE)
+        self.grant_type = kwargs.pop("grant_type", KEYCLOAK_GRANT_TYPE)
+        kwargs["prompt"] = kwargs.pop("prompt", "keycloak")
+
+        if KEYCLOAK_GRANT_TYPE == "authorization_code":
+          super().__init__(**kwargs)
+          log("keycloak auth: server_url=%s, client_id=%s, realm_name=%s, redirect_uri=%s, scope=%s, grant_type=%s",
+              self.server_url, self.client_id, self.realm_name, self.redirect_uri, self.scope, self.grant_type)
+
+          # Get authorization code
+          client = WebApplicationClient(KEYCLOAK_CLIENT_ID)
+          authorization_url = KEYCLOAK_SERVER_URL + 'realms/' + KEYCLOAK_REALM_NAME + '/protocol/openid-connect/auth'
+          self.salt = client.prepare_request_uri(
+                      authorization_url,
+                      redirect_uri = KEYCLOAK_REDIRECT_URI,
+                      scope = [KEYCLOAK_SCOPE],
+          )
+        else:
+          raise(NotImplementedError("Warning: only grant type \"authorization_code\" is currently supported."))
+
+    def __repr__(self):
+        return "keycloak"
+
+    def get_challenge(self, digests):
+        assert not self.challenge_sent
+        if "keycloak" not in digests:
+            log.error("Error: client does not support keycloak authentication")
+            return None
+        self.challenge_sent = True
+        return self.salt, "keycloak"
+
+    def check(self, response_json) -> bool:
+        assert self.challenge_sent
+        #log("response_json: %r", repr(response_json))
+        
+        if not response_json:
+          log.error("Error: keycloak authentication failed as invalid response was received from authorization endpoint.")
+          return False
+        
+        try:
+          response = json.loads(response_json)
+        except Exception:
+          log.error("Error: keycloak authentication failed as invalid response was received from authorization endpoint.")
```

Comment:
Please add:
```
log("json failed to load %r", response_json, exc_info=True)
```
So that the full details of the error can be shown when authentication debugging is enabled.
And perhaps split the long error message into two lines.

The exception type you should be catching here is [json.JSONDecodeError](https://docs.python.org/3/library/json.html#json.JSONDecodeError), not `Exception`.

Reply:
Do you want me to remove the `log.error()` and replace by 
```
log("json failed to load %r", response_json, exc_info=True)
```
, or keep both of them?

Reply:
Keep both of them.
`log.error` is for the more generic error message, `log` (which is an alias for `log.debug`) is only shown when debugging is enabled.